src/rust/multi.libgroup/template/src/template.rs

```diff
@@ -11,25 +13,77 @@ use aidoku::{
 use alloc::string::ToString;
 extern crate alloc;
 
-use crate::parser;
+use crate::{
+	helpers::{get_token, is_logged, save_token},
+	parser,
+};
 
 pub struct SocialLibSource {
 	pub site_id: &'static str,
 	pub domain: &'static str,
 	pub nsfw: &'static MangaContentRating,
-	pub cdn: &'static CDN,
 }
 
 pub struct CDN {
-	pub main: &'static str,
-	pub second: &'static str,
-	pub compress: &'static str,
+	pub main: String,
+	pub second: String,
+	pub compress: String,
 }
 
 static USER_AGENT: &str = "Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Mobile/15E148 Safari/604.1";
 static DOMAIN_API: &str = "https://api2.mangalib.me/api/";
 
 impl SocialLibSource {
+	pub fn refresh_token(&self) {
+		if is_logged() {
+			let timestamp = defaults_get("timestamp").unwrap().as_int().unwrap() / 1000;
+			let access_token = defaults_get("access_token")
+				.unwrap()
+				.as_string()
+				.unwrap()
+				.read();
+			let refresh_token = defaults_get("refresh_token")
+				.unwrap()
+				.as_string()
+				.unwrap()
+				.read();
+			let now = current_date() as i64;
+			let expires_in = defaults_get("expires_in").unwrap().as_int().unwrap();
+			if (now - timestamp) >= expires_in {
+				let url = format!("{}auth/oauth/token", DOMAIN_API);
+				let auth = format!("Bearer {}", access_token);
+				let domain = format!("https://{}/", self.domain);
+				let body = format!(
+					r#"{{"grant_type":"refresh_token","client_id":"{}","refresh_token":"{}","scope":""}}"#,
+					self.site_id, refresh_token
+				);
+
+				let request = Request::new(url, HttpMethod::Post)
+					.header("User-Agent", USER_AGENT)
+					.header("Site-Id", self.site_id)
+					.header("Content-Type", "application/json")
+					.header("Authorization", auth.as_str())
+					.header("Referer", domain.as_str())
+					.body(body);
+				let json = request.json().unwrap().as_object().unwrap();
+
+				save_token(json);
+			}
+		}
+	}
+
+	fn request_get(&self, url: &str) -> Request {
+		if is_logged() {
+			return Request::new(url, HttpMethod::Get)
+				.header("Site-Id", self.site_id)
+				.header("User-Agent", USER_AGENT)
+				.header("authorization", get_token().as_str());
+		}
+		Request::new(url, HttpMethod::Get)
+			.header("Site-Id", self.site_id)
+			.header("User-Agent", USER_AGENT)
```

Comment:
something like this reduces code duplication
```suggestion
		let req = Request::get(url)
			.header("Site-Id", self.site_id)
			.header("User-Agent", USER_AGENT);
		if is_logged() {
			req.header("authorization", get_token().as_str())
		} else {
			req
		}
```